### PR TITLE
add interfaces flag with unit test

### PR DESCRIPTION
**Pull request by kliukovkin**
This is a proposal for `-interfaces` flag using source mode. Using this flag it is possible to list only required interfaces to be mocked in source mode. Currently, source mode mocks all interfaces which -source file contains.

Thanks @KastenMike for raising this!

Fixes https://github.com/golang/mock/issues/660

### CODE REVIEW

File: mockgen/mockgen.go [hunk 1]
```diff
@@ -115,6 +116,14 @@ func main() {
 		return
 	}
 
+	if len(*interfaces) > 0 {
```
Comment (KastenMike) [inline mockgen/mockgen.go:119]: maybe this should go in `sourceMode` like `p.excludeNamesSet = parseExcludeInterfaces(*excludeInterfaces)`?

**Comment by kliukovkin** [comment]
Hey @JacobOaks , can you please take a look?

**Comment by JacobOaks** [comment]
Hey @kliukovkin! Thanks for the PR!

I am in favor of having this feature, just two questions/comments:
* Instead of using an `-interfaces` flag, what do you think about having source mode read positional arguments (i.e., `mockgen -source <path/to/file.go> InterfaceOne,InterfaceTwo`) to align it with how reflect mode works? (If there are no positional arguments specified, we would parse all interfaces to keep backwards compatibility)
* Instead of parsing and then dropping interfaces that aren't specified, can we simply not parse ones that aren't requested? This is similar to [how the exclusion flag already works](https://github.com/uber-go/mock/blob/main/mockgen/parse.go#L231-L233) and would avoid some wasted computation.

**Comment by kliukovkin** [comment]
> Hey @kliukovkin! Thanks for the PR!
> 
> 
> 
> I am in favor of having this feature, just two questions/comments:
> 
> * Instead of using an `-interfaces` flag, what do you think about having source mode read positional arguments (i.e., `mockgen -source <path/to/file.go> InterfaceOne,InterfaceTwo`) to align it with how reflect mode works? (If there are no positional arguments specified, we would parse all interfaces to keep backwards compatibility)
> 
> * Instead of parsing and then dropping interfaces that aren't specified, can we simply not parse ones that aren't requested? This is similar to [how the exclusion flag already works](https://github.com/uber-go/mock/blob/main/mockgen/parse.go#L231-L233) and would avoid some wasted computation.

Hey @JacobOaks ! Done, can you please review and let me know if that looks good?

**Comment by kliukovkin** [comment]
@JacobOaks hey Jacob! Gentle ping here.

File: mockgen/parse_test.go [hunk 2]
```diff
@@ -143,3 +146,127 @@ func TestParseArrayWithConstLength(t *testing.T) {
...
+						Name: "Bar",
+					},
+				},
+				requested: []string{"Foo", "Baz"},
+			},
+			want:    nil,
+			wantErr: true,
+		},
+		{
+			name: "missing interface (Baz not found)",
```
Comment (r-hang) [inline mockgen/parse_test.go:244]: Could we also add a test case for accidentally specified duplicate requested interfaces?

File: mockgen/parse_test.go [hunk 3]
```diff
@@ -143,3 +146,127 @@ func TestParseArrayWithConstLength(t *testing.T) {
...
+				requested: []string{"Baz"},
+			},
+			want:    nil,
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := filterInterfaces(tt.args.all, tt.args.requested)
+			if (err != nil) != tt.wantErr {
```
Comment (r-hang) [inline mockgen/parse_test.go:263]: instead of wantErr being a bool, could we make this a string and compare the error contents?

**Review by r-hang — APPROVED** [review]
Sorry for the delay. Looks good to me with some minor comments. Let's merge this when they are addressed.

File: mockgen/parse.go [hunk 4]
```diff
@@ -61,6 +63,20 @@ func sourceMode(source string) (*model.Package, error) {
...
+		p.includeNamesSet = make(map[string]struct{}, len(ifaces))
+		for _, arg := range ifaces {
+			for _, name := range strings.Split(arg, ",") {
+				name = strings.TrimSpace(name)
+				if name != "" {
+					p.includeNamesSet[name] = struct{}{}
+				}
+			}
+		}
+	}
```
Comment (abhinav) [inline mockgen/parse.go:78]: Other modes accept list of interfaces as a single argument, so I would recommend:

```suggestion
	if flag.NArg() > 1 {
		return nil, errors.New("-source mode accepts at most one argument")
	}
	if flag.NArg() == 1 {
		ifaces := strings.Split(flag.Arg(0), ",")
		p.includeNamesSet = make(map[string]struct{}, len(ifaces))
		for _, name := range ifaces {
			p.includeNamesSet[name] = struct{}{}
		}
	}
```

File: mockgen/parse.go [hunk 5]
```diff
@@ -168,6 +185,7 @@ type fileParser struct {
 	auxInterfaces      *interfaceCache
 	srcDir             string
 	excludeNamesSet    map[string]struct{}
+	includeNamesSet    map[string]struct{}
```
Comment (abhinav) [inline mockgen/parse.go:188]: worth adding a comment here:

```suggestion
	includeNamesSet    map[string]struct{} // empty to include all
```

File: mockgen/parse.go [hunk 6]
```diff
@@ -228,17 +246,36 @@ func (p *fileParser) parseFile(importPath string, file *ast.File) (*model.Packag
...
+		if _, ok := p.excludeNamesSet[name]; ok {
 			continue
 		}
-		i, err := p.parseInterface(ni.name.String(), importPath, ni)
+
+		if len(p.includeNamesSet) > 0 {
+			if _, ok := p.includeNamesSet[name]; !ok {
+				continue
+			}
+		}
```
Comment (abhinav) [inline mockgen/parse.go:259]:
```suggestion
		// All interfaces are included if no filter was specified.
		if len(p.includeNamesSet) > 0 {
			if _, ok := p.includeNamesSet[name]; !ok {
				continue
			}
		}
```

File: mockgen/parse.go [hunk 7]
```diff
@@ -228,17 +246,36 @@ func (p *fileParser) parseFile(importPath string, file *ast.File) (*model.Packag
...
 		is = append(is, i)
+
+		delete(p.includeNamesSet, name)
+	}
+	if len(p.includeNamesSet) > 0 {
+		missing := make([]string, 0, len(p.includeNamesSet))
+		for n := range p.includeNamesSet {
+			missing = append(missing, n)
+		}
+		sort.Strings(missing)
```
Comment (abhinav) [inline mockgen/parse.go:277]: We can shorten this:
```suggestion
		missing := slices.Sorted(maps.Keys))
```

(assuming import slices and maps.)

**Review by abhinav — APPROVED** [review]
@sywhang This PR implements the changes you suggested in #281 (failing on missing expected interfaces).

@kliukovkin, thanks for updating. left some minor suggestions for improvements.
I see that the tests are failing, though.

**Comment by kliukovkin** [comment]
Updated according to feedback.
- accepted all your suggestions
- removed the “missing interfaces” error check, because it was breaking integration tests - in practice, missing names should just be ignored

Thanks for the review @abhinav ! 🚀

**Review by sywhang — APPROVED** [review]
Thank you!
